_posts/2020-01-13-serverless-express-api.md

```diff
@@ -0,0 +1,562 @@
+      "author": "anonymous"
+    },
+    {
+      "message": "Message: 9",
+      "date": "2020-01-06T01:17:05.288Z",
+      "author": "anonymous"
+    },
+    {
+      "message": "Message: 8",
+      "date": "2020-01-06T01:17:04.876Z",
+      "author": "anonymous"
+    },
+    {
+      "message": "Message: 7",
+      "date": "2020-01-06T01:17:04.475Z",
+      "author": "anonymous"
+    },
+    {
+      "message": "Message: 6",
+      "date": "2020-01-06T01:17:04.070Z",
+      "author": "anonymous"
+    },
+    {
+      "message": "Message: 5",
+      "date": "2020-01-06T01:17:03.656Z",
+      "author": "anonymous"
+    },
+    {
+      "message": "Message: 4",
+      "date": "2020-01-06T01:17:02.156Z",
+      "author": "anonymous"
+    },
+    {
+      "message": "Message: 3",
+      "date": "2020-01-06T01:17:01.679Z",
+      "author": "anonymous"
+    },
+    {
+      "message": "Message: 2",
+      "date": "2020-01-06T01:17:00.576Z",
+      "author": "anonymous"
+    },
+    {
+      "message": "Message: 1",
+      "date": "2020-01-06T01:16:59.034Z",
+      "author": "anonymous"
+    }
+  ]
+}
+```
+
+Paging through the result set is possible by using the maxItems query parameter. If we set it to 1 e.g. `curl $ENDPOINT?maxItems=1`, we will receive the first item and a token to retrieve more data.
+
+```json
+{
+  "Items": [
+    {
+      "message": "Message: 10",
+      "date": "2020-01-06T01:17:05.691Z",
+      "author": "anonymous"
+    }
+  ],
+  "NextToken": "1mSTs9j4zPH1Dw7LkIZOXfUOpfd/vijAKI5gVyyeW0KjJ2BQqoxKgH3c2cyYQd74GYuEn0xQyKapxfmXVCd6yzT7cDUfA2pdExAWszRdL8EmEATzr4WMxeZQ5QtHJHpz7rN7q+8wIuE4mbwyYHCsBXf8ELJVmeRut"
+}
+```
+
+Now using the value of NextToken, we can retrieve the next value;
+
+```json
+{
+  "Items": [
+    {
+      "message": "Message: 9",
+      "date": "2020-01-06T01:17:05.288Z",
+      "author": "anonymous"
+    },
+  ],
+  "NextToken": "1XqRKzDWBlYIFrJLHMoTCTIHAjURZIAOz/Rq6EDIAihbhVcrXxV6Wzi5/WsNUvPeN1fCMVLUXgERX3w6zgQ7554S97HcGWZ+/iO2lkPj1JvGKTOd48u1qTdcywMCcmflcBR3oqd+aNROdH9nMa8KBIQHPSAfFy/SK"
+}
+```
+
+# A security perspective
+
+We now have a basic API working but there are still issues that remain.
+
+1. Our API is public, and all users are anonymous.
+
+   We haven't put in any authorization/authentication controls, which we would probably want to do in order to tie comments to a particular user. Adding an extra step to authenticate before-hand will likely reduce the number of potential calls to the API.
+
+2. We are not doing any validation of the incoming message.
+
+    There is a potential for the API to be used in an injection attacked. For example, a malicious javascript payload could be sent to the service and this could be executed by a browser upon retrieving a message. Whilst the client should ultimately be responsible for protecting against this, it would not be a bad idea to add some level of detection/sanitisation on the server side to make this less likely.
+
+# TODO
```

Comment:
Looking good. I haven't built it locally so not sure what it actually looks like..